### PR TITLE
[AIMIGRAPHX-885]  MLP tower batched horizontal fusions

### PR DESCRIPTION
**Pull request by TedThemistokleous**
Changes after further iteration/prototyping to ensure fusion doesn't break other horizontal passes, seeing a 4.8% improvment in lower batch model run performance

## Motivation

Customer model performance improvements as part of AIMIGRAPHX-885

Found a fusion opportunity with 4 identical MLP towers into batched GEMM chains. This uses horizontal fusion so that we can perform the 1 to N operation and combining prediction towers into a smaller subset of fused ops

The change here performs a chain aware fusion so that we can identify which fusion chains we can compile while also preserving other dot based fusions (vertical fusions on dot for eg) by using layer depth > 2 as a way to filter the chains we want to optimize.

The goal here is to reduce the amount of dots and fuse them with sigmoids, mul into one kernel and avoid a bunch of operations on smaller tensors via batching operations into a larger op which reduces overhead of launch scaling.

## Technical Details

Creates a fusion block to find, track (via hashing) and trace chains that capture a tower of MLP prediction layers. 

Compared to baseline develop I was seeing a 5% improvement on batch 1 reducing 24 individual dots and combining everything in the MLP chains (Sigmoid, Mul, Dot, Add) into a `mlir_dot_add_sigmoid_mul`  kernel.

The result is we take 24 operations into 3 Silu + 1 dot for these operations as it reduces 6 towers and 4 layers.

## Changelog Category

Add a `CHANGELOG.md` entry for any option other than `Not Applicable`
- - [x] Added: New functionality.
- - [ ] Changed: Changes to existing functionality.
- - [ ] Removed: Functionality or support that has been removed. (Compared to a previous release)
- - [ ] Optimized: Component performance that has been optimized or improved.
- - [ ] Resolved Issues: Known issues from a previous version that have been resolved.
- - [ ] Not Applicable: This PR is not to be included in the changelog.

### CODE REVIEW

File: src/fuse_horizontal.cpp [hunk 1]
```diff
@@ -267,24 +267,424 @@ struct gather_horizontal_fusion
...
+        }
+    }
+}
+
 void fuse_horizontal::apply(module_pass_manager& mpm) const
 {
     auto& m = mpm.get_module();
 
     fuse_horizontal_ops(m, gather_horizontal_fusion{});
+    fuse_mlp_chains(m);
```
Comment (pfultz2) [inline src/fuse_horizontal.cpp]: This should be extending the `fuse_horizontal_ops`, with another class.

File: src/fuse_horizontal.cpp [hunk 2]
```diff
@@ -267,24 +267,424 @@ struct gather_horizontal_fusion
...
+}
+
+// Trace an MLP chain forward from a root dot
+static mlp_chain trace_chain(instruction_ref first_dot)
+{
+    mlp_chain chain;
+    chain.root_input = first_dot->inputs().at(0);
+
+    auto current_dot = first_dot;
+    while(true)
```
Comment (pfultz2) [inline src/fuse_horizontal.cpp]: Dont use `while(true)` loops. I think `unfold` can be used traverse.

File: src/fuse_horizontal.cpp [hunk 3]
```diff
@@ -267,24 +267,424 @@ struct gather_horizontal_fusion
...
-//     bool is_candidate(instruction_ref ins) const { ... }
-//     std::string group_key(instruction_ref ins) const { ... }
-//     std::vector<instruction_ref>
-//         fuse(module& m, const std::vector<instruction_ref>& ops,
-//              instruction_ref insert_pt) const { ... }
-// };
+// Batches structurally-identical MLP chains end-to-end, keeping the batch
+// dimension through all layers including pointwise activations.
+//
+// Chain pattern:  dot → add(bias) → sigmoid → mul (SiLU)  → dot → ... → dot
```
Comment (pfultz2) [inline src/fuse_horizontal.cpp]: We can just fuse the `dot` operator in fuse_horizontal, and then fuse the add -> sigmoid -> mul in simplify_algebra. We dont need to fuse that here.

File: src/fuse_horizontal.cpp (hunk 1 above)
Comment (TedThemistokleous) [reply src/fuse_horizontal.cpp]: This is a prototype right now, putting this up in draft to work on for the rest of the short week. 

I know it needs cleanup. I'll mark it as ready for review. Its failing tidy and needs a few more things. I did however see a healthy boost on the lower batches for the customer model I was working on

File: src/fuse_horizontal.cpp (hunk 3 above)
Comment (TedThemistokleous) [reply src/fuse_horizontal.cpp]: Just curious, why? Is it to generalize the pipeline so that we can apply this for other chains like Gelu/Layernorm/other patterns?  Are you concerned about how things are encapsulated in fuse_horizontal here or is there some other flaw in the logic? 

More trying to understand the approach vs something a bit more targeting. The chain detection logic here is a bit more complex and tightly coupled. Do we typically want to split targetting chains like this between one simpler fusion and then match out the rest?

Simplify_algebra gets run with the optimize passes so I suppose I just throw a pass in the simplify_algebra , modify targets to add the optimize_module pass after fuse_horizontal to handle this?

Comment (TedThemistokleous) [reply src/fuse_horizontal.cpp]: Ah okay, so I think Ive got another changeset based on the ask. Tested it briefly this morning gave me similar performance boost but simplified a bunch of the logic but it looks like I can scale this potentially to other pieces/patterns too because if we generalize the dot and othrer pieces the rest can rely on the simplify algebra step. 

Threw tools at it for now. Will cleanup further tomorrow.

File: src/simplify_reshapes.cpp [hunk 4]
```diff
@@ -1811,6 +1811,28 @@ struct find_flatten
...
+        return match::any_of(unsqueeze(match::arg(0)(squeeze), same_shape_as_grandparent),
+                             squeeze(match::arg(0)(unsqueeze), same_shape_as_grandparent));
+    }
+
+    void apply(module& m, const match::matcher_result& mr) const
+    {
+        auto ins = mr.result;
+        m.replace_instruction(ins, ins->inputs().front()->inputs().front());
+    }
+};
```
Comment (pfultz2) [inline src/simplify_reshapes.cpp]: This should already be handled by `find_nested_shape_transforms`.

File: src/simplify_algebra.cpp [hunk 5]
```diff
@@ -2354,6 +2355,320 @@ struct find_conv_broadcast_input
...
+                auto new_slice = m.insert_instruction(
+                    std::next(lifted), csplits[i].slice->get_operator(), lifted);
+                auto new_squeeze = m.insert_instruction(
+                    std::next(new_slice), csplits[i].squeeze->get_operator(), new_slice);
+                m.replace_instruction(group[i], new_squeeze);
+            }
+            return;
+        }
+    }
+};
```
Comment (pfultz2) [inline src/simplify_algebra.cpp]: I think this can be much simpler. It should just match a `slice->squeeze->pw/reduce`(no need to look for splits) where the squeeze and slice have the same axis and rewrite it `slice->pw/reduce->squeeze` where you also unsqueeze the inputs to the pointwise/reduce op. The `find_op_shape_transform_op` in simplify_reshapes will take care of propagating it the rest of the way.

This also should go into simplify_reshapes.

File: src/fuse_horizontal.cpp [hunk 6]
```diff
@@ -267,24 +267,87 @@ struct gather_horizontal_fusion
...
+    bool is_candidate(instruction_ref ins) const
+    {
+        if(ins->name() != "dot")
+            return false;
+        if(ins->get_shape().dynamic())
+            return false;
+        if(ins->get_shape().lens().size() < 2)
+            return false;
+        auto weight = ins->inputs().at(1);
+        return weight->can_eval() and weight->get_shape().lens().size() >= 2;
```
Comment (pfultz2) [inline src/fuse_horizontal.cpp]: Why does the B matrix need to be const?

Also the check for ndim >= 2 seems redundant.

File: src/fuse_horizontal.cpp [hunk 7]
```diff
@@ -267,24 +267,87 @@ struct gather_horizontal_fusion
...
+{
+    std::size_t min_group_size() const { return 2; }
+
+    bool is_candidate(instruction_ref ins) const
+    {
+        if(ins->name() != "dot")
+            return false;
+        if(ins->get_shape().dynamic())
+            return false;
+        if(ins->get_shape().lens().size() < 2)
```
Comment (pfultz2) [inline src/fuse_horizontal.cpp]: Use `ndim()`.

File: src/fuse_horizontal.cpp [hunk 8]
```diff
@@ -267,24 +267,87 @@ struct gather_horizontal_fusion
...
+        std::vector<instruction_ref> results;
+        results.reserve(num);
+        for(std::size_t i = 0; i < num; ++i)
+        {
+            auto sliced = m.insert_instruction(
+                insert_pt,
+                make_op("slice",
+                        {{"axes", std::vector<int64_t>{0}},
+                         {"starts", std::vector<int64_t>{static_cast<int64_t>(i)}},
+                         {"ends", std::vector<int64_t>{static_cast<int64_t>(i + 1)}}}),
```
Comment (pfultz2) [inline src/fuse_horizontal.cpp]: Remove `static_cast`.

File: src/fuse_horizontal.cpp [hunk 9]
```diff
@@ -267,24 +267,87 @@ struct gather_horizontal_fusion
...
+            return false;
+        auto weight = ins->inputs().at(1);
+        return weight->can_eval() and weight->get_shape().lens().size() >= 2;
+    }
+
+    auto group_key(instruction_ref ins) const
+    {
+        return std::make_tuple(ins->inputs().at(0)->get_shape().lens(),
+                               ins->inputs().at(1)->get_shape().lens(),
+                               static_cast<int>(ins->get_shape().type()));
```
Comment (pfultz2) [inline src/fuse_horizontal.cpp]: remove `static_cast`.

File: src/fuse_horizontal.cpp (hunk 6 above)
Comment (TedThemistokleous) [reply src/fuse_horizontal.cpp]: There's an assumption that the weights are constant. In the current model I was using this for, this is the case. 

Should that not be the case. 

Yeah I can remove the >= 2 since we're bound by same_ndims() for our dot()

Comment (TedThemistokleous) [reply src/fuse_horizontal.cpp]: Removed the redundant dim check, kept the const in. Let me know if that's irrelevant here, or should be irrelevant. 

I'm leaning towards just turning this into `return true` so that its agnostic of which dot input is which

File: src/fuse_horizontal.cpp (hunk 7 above)
Comment (TedThemistokleous) [reply src/fuse_horizontal.cpp]: Fixed

File: src/fuse_horizontal.cpp (hunk 8 above)
Comment (TedThemistokleous) [reply src/fuse_horizontal.cpp]: Done

File: src/fuse_horizontal.cpp (hunk 9 above)
Comment (TedThemistokleous) [reply src/fuse_horizontal.cpp]: Done

File: src/simplify_algebra.cpp (hunk 5 above)
Comment (TedThemistokleous) [reply src/simplify_algebra.cpp]: As in move everything to simplify_reshapes, and remove this entirely? Okay, let me try that and get back to you. 

Quick note and question - I guess anything related to any sort or rewriting or triggering a certain fusion should be put into simplify_reshapes then?

File: src/simplify_reshapes.cpp (hunk 4 above)
Comment (TedThemistokleous) [reply src/simplify_reshapes.cpp]: Let me try and rewrite.

File: src/fuse_horizontal.cpp (hunk 6 above)
Comment (TedThemistokleous) [reply src/fuse_horizontal.cpp]: Nah can't just make this true beacuse we assume its data, weight, unless we should be checking if its xor (one is const, one is non const) here so that we're only fusing constant weights and don't make an assumption about ordering.

File: src/simplify_reshapes.cpp (hunk 4 above)
Comment (TedThemistokleous) [reply src/simplify_reshapes.cpp]: Was able to just remove this. Thanks for the heads up.

File: src/simplify_algebra.cpp (hunk 5 above)
Comment (pfultz2) [reply src/simplify_algebra.cpp]: If its related to reshape/transpose/broadcast then usually we put this in simplify_reshapes(althougth the horizontal fusions of reshape/transpose are done in simplify_algebra just because that is where the `get_splits` function is, but I would like to move it to simplify_reshapes in the future).

File: src/simplify_reshapes.cpp [hunk 10]
```diff
@@ -1811,6 +1811,204 @@ struct find_flatten
...
+// Match slice->squeeze->pw/reduce where the squeeze and slice share the same
+// single axis, then rewrite to slice->pw/reduce->squeeze (unsqueezing the
+// other inputs).  find_op_shape_transform_op propagates the squeeze through
+// any downstream op chain, and find_splits in simplify_algebra merges parallel
+// branches back together.
+struct find_slice_squeeze
+{
+    auto matcher() const
+    {
+        auto match_op = match::any_of(match::name("pointwise"), match::pointwise(), match::reduce());
```
Comment (pfultz2) [inline src/simplify_reshapes.cpp]: `match::name("pointwise")` is redundant here.

File: src/simplify_reshapes.cpp [hunk 11]
```diff
@@ -1811,6 +1811,204 @@ struct find_flatten
...
+        {
+            if(input == squeeze)
+                input = slice_ins;
+            else
+                input = m.insert_instruction(
+                    op_ins, make_op("unsqueeze", {{"axes", {axis}}}), input);
+        }
+
+        auto op = op_ins->get_operator();
+        if(not op.attributes().contains("pointwise") and op_ins->name() != "pointwise")
```
Comment (pfultz2) [inline src/simplify_reshapes.cpp]: `op_ins->name() != "pointwise"` is redundant.

File: src/simplify_algebra.cpp [hunk 12]
```diff
@@ -38,6 +38,7 @@
 #include <migraphx/instruction.hpp>
 
 #include <migraphx/algorithm.hpp>
+#include <optional>
```
Comment (pfultz2) [inline src/simplify_algebra.cpp]: This can be removed.

File: src/simplify_reshapes.cpp [hunk 13]
```diff
@@ -1811,6 +1811,204 @@ struct find_flatten
...
+
+                v["axes"] = op_axes;
+                op = make_op(op_ins->name(), v);
+            }
+            else if(v.contains("axis"))
+            {
+                auto a = v["axis"].to<int64_t>();
+                if(a >= axis)
+                    a++;
+                v["axis"] = a;
```
Comment (pfultz2) [inline src/simplify_reshapes.cpp]:
```suggestion
                v["axis"] = a >= axis ? a + 1 : a;
```

**Review by pfultz2 — CHANGES_REQUESTED** [review]
I would like to see a a perf report ran before we merge this in to make sure this doesnt cause a perf regression in other models. I know in the past when I have done changes like this, there is usually some other changes needed to avoid regressions in other models.

**Comment by TedThemistokleous** [comment]
> I would like to see a a perf report ran before we merge this in to make sure this doesnt cause a perf regression in other models. I know in the past when I have done changes like this, there is usually some other changes needed to avoid regressions in other models.

Can do. I actually did this to track progress. Do you want it on the ticket or in the model review?  Also would you like that on the other PRs under the similar tag (AIMIGRAPHX-885). They all modify the summary in specific ways

File: src/simplify_reshapes.cpp (hunk 10 above)
Comment (TedThemistokleous) [reply src/simplify_reshapes.cpp]: Removed

File: src/simplify_algebra.cpp (hunk 12 above)
Comment (TedThemistokleous) [reply src/simplify_algebra.cpp]: Removed

File: src/simplify_reshapes.cpp (hunk 13 above)
Comment (TedThemistokleous) [reply src/simplify_reshapes.cpp]: Done

File: src/simplify_reshapes.cpp (hunk 11 above)
Comment (TedThemistokleous) [reply src/simplify_reshapes.cpp]: Done

File: src/simplify_reshapes.cpp [hunk 14]
```diff
@@ -1812,6 +1812,202 @@ struct find_flatten
...
+        m.replace_instruction(op_ins, new_sq);
+    }
+};
+
+// Match squeeze(pw_dag(slice)) where every non-constant leaf of the pointwise
+// DAG is the same slice instruction and squeeze/slice share the same single
+// axis.  Lifts the entire DAG above the slice:
+//   slice -> pw_dag -> squeeze   =>   pw_dag(root) -> slice -> squeeze
+// Each branch produces its own copy; CSE merges duplicates later.
+struct find_slice_pw_subgraph
```
Comment (pfultz2) [inline src/simplify_reshapes.cpp]: This should be removed as its already handled by `find_splits` in simplify_algebra.

File: test/simplify_reshapes_test.cpp [hunk 15]
```diff
@@ -45,6 +46,20 @@ static void run_pass(migraphx::module& m)
...
+    {
+        migraphx::run_passes(m,
+                             {migraphx::simplify_reshapes{.enable_op_shape_transform_op = true},
+                              migraphx::dead_code_elimination{},
+                              migraphx::simplify_algebra{},
+                              migraphx::dead_code_elimination{}});
+    }
+    migraphx::run_passes(
+        m, {migraphx::eliminate_common_subexpression{}, migraphx::dead_code_elimination{}});
+}
```
Comment (pfultz2) [inline test/simplify_reshapes_test.cpp]: Dont do this here. If you need unit tests across multiple passes, you can add a unit test to optimize_module_test.cpp, which runs optimize_module.

File: src/fuse_horizontal.cpp [hunk 16]
```diff
@@ -267,24 +267,87 @@ struct gather_horizontal_fusion
...
+            return false;
+        auto weight = ins->inputs().at(1);
+        return weight->can_eval();
+    }
+
+    auto group_key(instruction_ref ins) const
+    {
+        return std::make_tuple(ins->inputs().at(0)->get_shape().lens(),
+                               ins->inputs().at(1)->get_shape().lens(),
+                               static_cast<int>(ins->get_shape().type()));
```
Comment (bdevorem) [inline src/fuse_horizontal.cpp]: is this static cast needed?

File: src/simplify_reshapes.cpp [hunk 17]
```diff
@@ -1812,6 +1812,202 @@ struct find_flatten
...
+                    op_ins, make_op("unsqueeze", {{"axes", {axis}}}), input);
+        }
+
+        auto op = op_ins->get_operator();
+        if(not op.attributes().contains("pointwise"))
+        {
+            auto v = op.to_value();
+            if(v.contains("axes"))
+            {
+                auto op_axes = v["axes"].to_vector<int64_t>();
```
Comment (bdevorem) [inline src/simplify_reshapes.cpp:1920]: are negative axes already normalized at this point? I think so but just sanity checking

File: test/simplify_reshapes_test.cpp [hunk 18]
```diff
@@ -5040,6 +5055,339 @@ TEST_CASE(gather_strided_view_elements_mismatch)
...
+TEST_CASE(slice_squeeze_reduce_sum)
+{
+    migraphx::shape s{migraphx::shape::float_type, {2, 4, 8}};
+    migraphx::module m1;
+    {
+        auto input = m1.add_parameter("input", s);
+        auto sl    = m1.add_instruction(
+            migraphx::make_op("slice", {{"axes", {0}}, {"starts", {0}}, {"ends", {1}}}), input);
+        auto sq  = m1.add_instruction(migraphx::make_op("squeeze", {{"axes", {0}}}), sl);
+        auto red = m1.add_instruction(migraphx::make_op("reduce_sum", {{"axes", {1}}}), sq);
```
Comment (bdevorem) [inline test/simplify_reshapes_test.cpp]: can you add a test where the reduce is over all axes?

File: test/simplify_reshapes_test.cpp [hunk 19]
```diff
@@ -5040,6 +5055,339 @@ TEST_CASE(gather_strided_view_elements_mismatch)
...
+{
+    migraphx::shape s{migraphx::shape::float_type, {2, 4}};
+    migraphx::module m1;
+    {
+        auto input = m1.add_parameter("input", s);
+        auto s0    = m1.add_instruction(
+            migraphx::make_op("slice", {{"axes", {0}}, {"starts", {0}}, {"ends", {1}}}), input);
+        auto sq0 = m1.add_instruction(migraphx::make_op("squeeze", {{"axes", {0}}}), s0);
+        auto s1  = m1.add_instruction(
+            migraphx::make_op("slice", {{"axes", {0}}, {"starts", {1}}, {"ends", {2}}}), input);
```
Comment (bdevorem) [inline test/simplify_reshapes_test.cpp]: can you add a test where the two inputs to the pw are from different squeeze/slice chains?

**Review by bdevorem — CHANGES_REQUESTED** [review]
there's a couple other tests I think would be good for corner case coverage but otherwise lgtm

File: src/fuse_horizontal.cpp (hunk 16 above)
Comment (TedThemistokleous) [reply src/fuse_horizontal.cpp]: I'll remove

File: src/simplify_reshapes.cpp (hunk 17 above)
Comment (TedThemistokleous) [reply src/simplify_reshapes.cpp:1920]: I can double check this . Good catch

Comment (TedThemistokleous) [reply src/simplify_reshapes.cpp:1920]: This is needed not normalized

File: test/simplify_reshapes_test.cpp (hunk 18 above)
Comment (TedThemistokleous) [reply test/simplify_reshapes_test.cpp]: Done

File: test/simplify_reshapes_test.cpp (hunk 19 above)
Comment (TedThemistokleous) [reply test/simplify_reshapes_test.cpp]: Done

File: test/simplify_reshapes_test.cpp (hunk 15 above)
Comment (TedThemistokleous) [reply test/simplify_reshapes_test.cpp]: Done, should be pushing a commit that removes this and pushes tests that use this logic to optimize_module_test

File: src/simplify_reshapes.cpp (hunk 14 above)
Comment (TedThemistokleous) [reply src/simplify_reshapes.cpp]: removed

**Comment by TedThemistokleous** [comment]
> I would like to see a a perf report ran before we merge this in to make sure this doesnt cause a perf regression in other models. I know in the past when I have done changes like this, there is usually some other changes needed to avoid regressions in other models.

Got to the end of the line with this oen I think, I'm getting around a 2% boost after its all said and done...analyzing this further after changes from develop we're losing the `mlir_slice_sigmoid_mul_dot_add_unsqueeze` kernels. Did some more effort to add that back and making this a bit more robust gives us a small boost, and actually eliminates things further but the gain seems marginal.

Seeing develop going from 230-> 237 QPS and it adds a newly created `mlir_dot_add_sigmoid_mul_squeeze`

This ended up removing a 16 mlir_dot_add_sigmoid_mul kernels into 6 of the `mlir_dot_add_sigmoid_mul_squeeze`

Not sure if the effort here is worth the payoff - changes need more cleanup. we can discuss this further if you're free tomorrow/friday. I kept a bunch if intermediate perf IR reports if you wanna go over these

**Comment by TedThemistokleous** [comment]
@pfultz2  following up on this, without the --enable-offload-copy, we do see a perf boost so this isn't wasted work. Trying to tackle overhead right now as the offload copy adds so much that it overshadows any perf gains we see when using this on lower batch sizes. With the offload copy we get 2% but thats including hipMemCpy that take upwards of 50-60% of the run.

File: src/fuse_horizontal.cpp [hunk 20]
```diff
@@ -267,24 +267,242 @@ struct gather_horizontal_fusion
...
+
+    struct silu_detect_result
+    {
+        std::vector<instruction_ref> sig_insts;
+        std::vector<instruction_ref> mul_insts;
+    };
+
+    // Check if every dot feeds into add(dot, broadcast(...)). If so, return
+    // the add instructions and bias broadcast instructions.
+    static bool detect_downstream_biases(const std::vector<instruction_ref>& dots,
```
Comment (pfultz2) [inline src/fuse_horizontal.cpp:361]: This shouldnt be fusing the downstream elementwise operators. That should be handled by the other horizontal fusions we have.

File: src/simplify_reshapes.cpp [hunk 21]
```diff
@@ -976,6 +976,65 @@ struct find_concat_reshape
     }
 };
 
+// find_concat_reshape cannot handle the case where unsqueeze inserts a new
```
Comment (pfultz2) [inline src/simplify_reshapes.cpp:979]: It seems like we should update `find_concat_reshape` to handle those cases. We could use the shape_transform_descriptor to help with mapping the axes and rebasing the reshape.

**Comment by pfultz2** [comment]
I dont if maybe this should be split into 3 PRs, one for `dot_horizontal_fusion`, `find_concat_reshape` and another for `find_slice_squeeze`. At least the `find_concat_reshape` might need to be a PR on its own.

**Comment by TedThemistokleous** [comment]
> I dont if maybe this should be split into 3 PRs, one for `dot_horizontal_fusion`, `find_concat_reshape` and another for `find_slice_squeeze`. At least the `find_concat_reshape` might need to be a PR on its own.

Not opposed to that, there's actually more changes ontop of this that added some additional changes until I hit a dead end I haven't pushed up to this PR. Likely need to get cleaned up again. 

The bigger issue I believe is these updates that do fix and give a smaller boost (7% without offload copy) are completely overshadowed by the copies resulting really not much of a difference.  Ive got another changeset Ill introduce once I'm done testing that handle some additional overhead with Context as I was seeing memory idle gaps in our execution pipeline between kernels for the lower batches.

File: src/simplify_reshapes.cpp (hunk 21 above)
Comment (pfultz2) [reply src/simplify_reshapes.cpp:979]: Actually, I think this might be ok for now,

Comment (TedThemistokleous) [reply src/simplify_reshapes.cpp:979]: Okay resolving this then.

File: src/fuse_horizontal.cpp (hunk 20 above)
Comment (TedThemistokleous) [reply src/fuse_horizontal.cpp:361]: Which one? RIght now this gives me an 11% perf boost, but since the model I'm runnign this on has a copy_from_gpu after this there's a bit of a slowdown till I do something on the ORT side for memory coalescing for the outputs. This shows up when we do --enable-offload-copy too.

Comment (pfultz2) [reply src/fuse_horizontal.cpp:361]: > Which one?

`find_splits` in simplify_algebra.

File: src/simplify_reshapes.cpp [hunk 22]
```diff
@@ -1808,6 +1867,341 @@ struct find_flatten
...
+        }
+        else
+        {
+            break;
+        }
+    }
+    return {sig, current};
+}
+
+static void hoist_pointwise_above_slices(module& m)
```
Comment (TedThemistokleous) [inline src/simplify_reshapes.cpp:2054]: Delete

File: src/simplify_reshapes.cpp [hunk 23]
```diff
@@ -1808,6 +1867,341 @@ struct find_flatten
...
+    return true;
+}
+
+// Walk a pointwise chain from `start` and return the signature plus
+// the terminal instruction.  Handles both linear chains (single
+// consumer) and diamond patterns where `current` has exactly two
+// outputs that converge to a single pointwise merge (e.g. SiLU:
+// slice → sigmoid → mul(slice, sigmoid)).
+static std::pair<pw_chain_sig, instruction_ref>
+trace_pw_chain(instruction_ref start)
```
Comment (TedThemistokleous) [inline src/simplify_reshapes.cpp:1973]: Delete

**Comment by TedThemistokleous** [comment]
concat_unsqueeze - https://github.com/ROCm/AMDMIGraphX/pull/4984
Two others to come

**Comment by TedThemistokleous** [comment]
closing as this is split out as #5014 #5004 #4984
